src/ui/public/vislib/lib/tilemap_settings.js

```diff
@@ -0,0 +1,135 @@
+      _url: mapsConfig.deprecated.config.url,
+      _options: optionsFromConfig,
+
+      /**
+       * Must be called before getUrl/getOptions can be called.
+       */
+      whenSettingsReady: async function () {
+
+        if (!mapsConfig.deprecated.isConfiguredWithDefault) {//if settings are overridden, we will use those.
+          return true;
+        }
+
+        if (this._settingsInitialized) {
+          return true;
+        }
+
+        let manifest;
+        try {
+          const response = await getTileServiceManifest(mapsConfig.config.manifest, this._queryParams, attributionFromConfig, optionsFromConfig);
+          manifest = response.data;
+        } catch (e) {
+          //request failed. Continue to use old settings.
+          this._settingsInitialized = true;
+          return true;
+        }
+
+        this._options = {
+          attribution: $sanitize(marked(manifest.services[0].attribution)),
+          minZoom: manifest.services[0].minZoom,
+          maxZoom: manifest.services[0].maxZoom,
+          subdomains: []
+        };
+
+        const queryparams = _.assign({license: this._licenseUid}, manifest.services[0].query_parameters);
+        const query = url.format({query: queryparams});
+        this._url = manifest.services[0].url + query;//must preserve {} patterns from the url, so do not format path.
+
+        this._settingsInitialized = true;
+        return true;
+      },
+      /**
+       * add optional query-parameter that will be submitted to the tile service
+       * @param additionalQueryParams
+       */
+      addQueryParams: function (additionalQueryParams) {
+        this._queryParams = _.assign({}, this._queryParams, additionalQueryParams);
+      },
+
+      /**
+       * Get the url of the default TMS
+       * @return {string}
+       */
+      getUrl() {
+        return this._url;
+      },
+      /**
+       * Get the options of the default TMS
+       * @return {{}}
+       */
+      getOptions() {
+        return this._options;
+      }
+    };
+
+    return mapSettings;
+
+
+    /**
+     * mocks call to manifest
+     * will be replaced with AJAX-call to the service when available.
+     *       A sample response would look something like;
+     *
+     {
+   "version":"0.0.0",
+   "expiry":"14d",
+   "services":[
+      {
+         "id":"road_map",
+         "url":"https://proxy-tiles.elastic.co/v1/default/{z}/{x}/{y}.png",
+         "minZoom":0,
+         "maxZoom":12,
+         "attribution":"© [Elastic Tile Service](https://www.elastic.co/elastic-tile-service)",
+         "query_parameters":{
+            "elastic_tile_service_tos":"agree",
+            "my_app_name":"kibana"
+         }
+      }
+   ]
+    }
+     *
+     */
+    async function getTileServiceManifest(manifestUrl, additionalQueryParams) {
+
+      const manifestServiceTokens = url.parse(manifestUrl);
+      manifestServiceTokens.query = _.assign({}, manifestServiceTokens.query, additionalQueryParams);
```

Comment:
Unfortunately `url.parse` is horrible. You have to pass `true` for the second arg of `url.parse` in order for the query string to be parsed, and then when formatting it it's overridden by the `search` property of the parsed url anyway.

I suggest using a helper like [this one](https://github.com/elastic/kibana/blob/master/test/support/page_objects/common.js#L58-L65) from the function tests.